src/Servers/Kestrel/Core/src/KestrelServer.cs

```diff
@@ -211,73 +211,135 @@ public async Task StopAsync(CancellationToken cancellationToken)
 
-                for (int i = connectionTransportCount; i < totalTransportCount; i++)
+                IChangeToken reloadToken = null;
+
+                if (Options.ConfigurationLoader?.ReloadOnChange == true && (!_serverAddresses.PreferHostingUrls || _serverAddresses.Addresses.Count == 0))
                 {
-                    (IMultiplexedConnectionListener listener, Task acceptLoop) = _multiplexedTransports[i - connectionTransportCount];
-                    tasks[i] = Task.WhenAll(listener.UnbindAsync(cancellationToken).AsTask(), acceptLoop);
+                    reloadToken = Options.ConfigurationLoader.Configuration.GetReloadToken();
                 }
 
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+                Options.ConfigurationLoader?.Load();
 
-                if (!await ConnectionManager.CloseAllConnectionsAsync(cancellationToken).ConfigureAwait(false))
-                {
-                    Trace.NotAllConnectionsClosedGracefully();
+                await AddressBinder.BindAsync(Options.ListenOptions, AddressBindContext).ConfigureAwait(false);
+                _configChangedRegistration = reloadToken?.RegisterChangeCallback(async state => await ((KestrelServer)state).RebindAsync(), this);
+            }
+            finally
+            {
+                _bindSemaphore.Release();
+            }
+        }
 
-                    if (!await ConnectionManager.AbortAllConnectionsAsync().ConfigureAwait(false))
-                    {
-                        Trace.NotAllConnectionsAborted();
-                    }
-                }
+        private async Task RebindAsync()
+        {
+            await _bindSemaphore.WaitAsync();
+
+            IChangeToken reloadToken = null;
 
-                for (int i = 0; i < connectionTransportCount; i++)
+            try
+            {
+                if (_stopping == 1)
                 {
-                    (IConnectionListener listener, Task acceptLoop) = _transports[i];
-                    tasks[i] = listener.DisposeAsync().AsTask();
+                    return;
                 }
 
-                for (int i = connectionTransportCount; i < totalTransportCount; i++)
+                reloadToken = Options.ConfigurationLoader.Configuration.GetReloadToken();
+                var (endpointsToStop, endpointsToStart) = Options.ConfigurationLoader.Reload();
+
+                if (endpointsToStop.Count > 0)
                 {
-                    (IMultiplexedConnectionListener listener, Task acceptLoop) = _multiplexedTransports[i - connectionTransportCount];
-                    tasks[i] = listener.DisposeAsync().AsTask();
+                    var urlsToStop = endpointsToStop.Select(lo => lo.EndpointConfig.Url ?? "<unknown>");
+                    Trace.LogInformation("Config changed. Stopping the following endpoints: {endpoints}", string.Join(", ", urlsToStop));
+
+                    // 5 is the default value for WebHost's "shutdownTimeoutSeconds", so use that.
```

Comment:
Review: should we use the actual shutdownTimeoutSeconds value? Is that reasonably trivial to access?

Reply:
Probably not, this is a different situation. Endpoint should stop pretty quickly if they're just stopping the listener, no? Not like we need to wait for requests to drain?